### DISCUSSION TITLE
DOCS: Realtime with custom JWT

### DISCUSSION THREAD

**Discussion by tg44** [supabase/supabase#11826] [Questions]
# Improve documentation

## Link 

[Replication page](https://supabase.com/docs/guides/database/replication)
or
[Realtime API](https://supabase.com/docs/guides/api#realtime-api-1)

## Describe the problem

The custom JWT part is missing.

## Describe the improvement

I would add a section which at least mentions that you can do some realtime with a custom JWT, but it will only work if the claim contains a `role` field which should be a valid db role eg `role: "authenticated"`.
Also, I would move out this section to a separate page (or split out the whole page to multiple smaller pages), and would add some debugging ideas, like you can check the subscriptions in the realtime schema, and if you not seeing any rows, it is probably a misconfiguration. Or to check the wss messages in the console. Or how to check if even realtime is enabled for the table.

Splitted out from #7339

**Comment by kiwicopple** [comment]
Thanks @tg44 !

just to make sure that I'm clear on the audience we are writing for, what was the use-case for you here? Were you using a different auth system?

**Comment by tg44** [comment]
We are using web3 token access.

**Comment by cioionut** [comment]
Hi,
In my case, the subscription is created in `realtime.subscription` table but it disappears after a few seconds, and while the subscription record is present the stream works, and then it stops.

I use the latest version of [supabase_flutter](https://pub.dev/packages/supabase_flutter)
And to receive the stream values I use this https://supabase.com/docs/reference/dart/stream

It is worth mentioning that I sign the JWT using the supabase jwt secret, and the payload contains:
`{
  "supabase_uid": "1ed9dcf5-e383-4e41-820f-58df106c3e5c",
  "role": "authenticated",
  "aud": "authenticated",
  "app_metadata": {},
  "user_metadata": {},
  "iss": "my_supabase",
  "iat": 1663875937,
  "exp": 1663880737
}`

and then I use `supabase.auth.setAuth(token)` to set the credentials. Even when I set the auth with a token that was generated by supabase auth the subscriptions vanish.

When I use the logs explorer I see the following errors:
"Received an invalid access token from client: {:error, :unknown}"
"Unknown connection authorization error: :error"

The same errors occurred when I `setAuth(with_jwt_stolen_from_signin_session)`.
The subscriptions seem to work well when I sign in using the client SDK, but this is not what I want I need to use my authentication system. In all other cases (excepting streams) when I call the database using my JWT the results are ok.

PS: My RLS policies require just an authenticated role. And the replication is ON for that table.
Are there any mentions about how supabase authorizes realtime requests in order to avoid this kind of issue?

Edit: I test my code on a self-hosted version and it works as expected. But continues to terminate the subscriptions after a few seconds when I use Supabase online platform.

**Comment by GaryAustin1** [comment]
@cioionut check this out regarding a flutter issue that sound like your problem.  https://github.com/supabase/supabase/discussions/9146

**Comment by cioionut** [comment]
I see, thank you for your help but the issues with the authorization persists.

With `supabase_flutter: ^1.0.0-dev.9` I get the following behavior:
1. My subscriptions are now present in `realtime.subscription`. (Subscription vanish issue is gone).
2. But instead of one subscription per
 `supabase.from('messages').stream(['id']).eq('convo_id', chatId).order('created_at');`
I get three entries in the subscription table with `anon` claims instead of `authenticated` claims as expected since I call `supabase.auth.setAuth(myToken)`. The token was created using the settings described in my last comment.
2.1. In the `subscription` table the column `claims` contains: `{"exp":1977610453,"iat":1662034453,"iss":"supabase","ref":"<my_project_id>","role":"anon"}`
2.2. In the `subscription` table the column `claims_role` contains `anon`.
3. The stream is working well when RLS Policy is disabled, but it stops when enable with role `authenticated` (I assume because of the permissions, as I said at 2. my subscriptions have a claim role anon).

The logs for realtime are empty and I can't see any clues.

When I try in a self-hosted environment (using docker image: `supabase/realtime:v0.24.2`) I get one single subscription per `.stream` call but the `anon` claims persist in the `realtime.subscription` entry and it works when RLS are disabled.

PS: With docker image: `supabase/realtime:v0.22.4` and `supabase_flutter:0.3.3` the stream work as expected.

**Comment by GaryAustin1** [comment]
@cioionut Without going to check the realtime code my guess is because you don't have sub (usually set to user UUID) in your jwt for authenticated it is messing up.

**Comment by cioionut** [comment]
I've tried with a valid `sub` but no results.

1. My question is why do I need to set the `sub == user_id` (from auth)?  Does this mean that I can no longer use other authentication system? 
2. What about `session_id` is this claim also a requirement?

Until now I found that in order to make authorized calls to supabase database the custom JWT needs to have in it the following claims:
`role`, `aud`, `sub`, `iat`, `exp`. Is correct? And optional, other custom claims to check against in RLS policies.

I will try to make a PR to update the docs when I will find out what are the requirements for a custom JWT to work in realtime.

**Comment by paddyfink** [comment]
Hello,
I got a similar issue and the realtime feature is not working when I use a custom token.
My decoded token is: 
```
{
  "aud": "authenticated",
  "sub": "abc123",
  "iat": 1673878423,
  "exp": 1673964823,
  "email": "test@email.com",
  "name": "test",
  "role": "authenticated"
}
```
There is a claim missing in order for the realtime to work?

**Comment by hf** [comment]
Hey everyone, Realtime uses Postgres RLS to identify users. If you are minting custom JWT tokens your RLS policies need to be compatible with them. 

Usually this means that you must have a `sub` claim that is a UUID of a user record in the `auth.users` table.

**Reply by paddyfink** [reply]
@hf the whole point of custom jwt tokens  is to be able to use supabase without the auth module. like https://supabase.com/docs/guides/integrations/auth0

So what you are saying is if we integrate supabase with another identity provider, we lose the realtime feature?

**Reply by hf** [reply]
> @hf the whole point of custom jwt tokens  is to be able to use supabase without the auth module. like https://supabase.com/docs/guides/integrations/auth0
> 
> So what you are saying is if we integrate supabase with another identity provider, we lose the realtime feature?

I'm saying that if your RLS policies look up users in the Supabase managed `auth.users` table, you have to format the JWT properly as the `auth.uid()` function uses the `sub` claim.

You also need to make sure you're setting the `aud` claim to `authenticated` if your RLS policies use `to authenticated`.

**Reply by tg44** [reply]
https://github.com/supabase/supabase/issues/7339 something like this should work without using any kind of (valid) sub, am I right?

**Reply by hf** [reply]
It really depends on what your RLS policies use. I believe the only requirements are to have an `aud` and `role` claim. These claims are used to "assume" the database role so that realtime can correctly execute RLS policies.

If you use `auth.uid()` in your RLS policies there has to be a `sub` claim that is a valid UUID.

We definitely should improve the docs on this.

**Comment by hf** [comment]
I'm thinking this issue is more like a discussion than a bug or something that needs fixing, so I'll convert it to a discussion.

**Reply by tg44** [reply]
As the whole thing starts; its a documentation improvement claim in this area. We should not have a discussion about this, it should be written (or at least mentioned properly) in the docs. I used the doc improve issue template back then.

**Comment by w3b6x9** [comment]
I added instructions on how to use custom tokens with `supabase-js` in the docs here: https://supabase.com/docs/guides/realtime/postgres-changes#custom-tokens

**Reply by tg44** [reply]
I'm not sure this solves my original problem.

What I missed from the documentation;
 - If I download the jwt secret from the admin panel, what should I do to create a JWT?
 - How can I write an RLS role that handles the custom JWT?
 - What should I include to the JWT to make it work with realtime?
 - What should I check to make the debugging process less stressful?
 
 When I used the custom jwt method almost a year ago I needed to collect all of these informations from random issues and SO questions. Your mod in the docs not really answers any of these questions.
 
 BTW [this](https://supabase.com/docs/learn/auth-deep-dive/auth-deep-dive-jwts) and [this](https://supabase.com/docs/learn/auth-deep-dive/auth-policies) get much better.

**Reply by tg44** [reply]
And here I am again, and still need to dig up things like how can I add custom fields to the jwt...

**Reply by WesleyYue** [reply]
What about tokens that need to be refreshed? Are we supposed to unsubscribe, create a new client with new token, and re-subscribe?

**Reply by nathantew14** [reply]
@tg44 thank you, this is my exact situation as well.
